imports/server/guesses.ts

```diff
@@ -57,6 +62,167 @@ function transitionGuess(guess: GuessType, newState: GuessType['state']) {
+        this.exists = true;
+        this.sub.added(model.tableName, id, fields);
+      },
+      changed: (_, fields) => {
+        this.sub.changed(model.tableName, id, fields);
+      },
+      removed: () => {
+        this.exists = false;
+        this.sub.removed(model.tableName, id);
+      },
+    });
+  }
+
+  destroy() {
+    if (this.exists) {
+      this.sub.removed(this.tableName, this.id);
+    }
+
+    this.handle.stop();
+  }
+}
+
+class RefCountedObserverMap<T extends BaseType> {
+  private sub: Subscription;
+
+  private model: Base<T>;
+
+  private subscribers: Map<string, { refCount: number, subscriber: RecordUpdateObserver<T> }>;
+
+  constructor(sub: Subscription, model: Base<T>) {
+    this.sub = sub;
+    this.model = model;
+    this.subscribers = new Map();
+  }
+
+  incref(id: string) {
+    if (this.subscribers.has(id)) {
+      this.subscribers.get(id)!.refCount += 1;
+    } else {
+      this.subscribers.set(id, {
+        refCount: 1,
+        subscriber: new RecordUpdateObserver(this.sub, id, this.model),
+      });
+    }
+  }
+
+  decref(id: string) {
+    const record = this.subscribers.get(id);
+    if (!record) {
+      return;
+    }
+    record.refCount -= 1;
+    if (record.refCount <= 0) {
+      record.subscriber.destroy();
+      this.subscribers.delete(id);
+    }
+  }
+}
+
+class PendingGuessWatcher {
+  sub: Subscription
+
+  guessCursor: Mongo.Cursor<GuessType>
+
+  guessWatch: Meteor.LiveQueryHandle
+
+  guesses: Record<string, GuessType>
+
+  huntRefCounter: RefCountedObserverMap<HuntType>;
+
+  puzzleRefCounter: RefCountedObserverMap<PuzzleType>
+
+  constructor(sub: Subscription) {
+    this.sub = sub;
+
+    this.guessCursor = Guesses.find({ state: 'pending' });
+    this.guesses = {};
+    this.huntRefCounter = new RefCountedObserverMap(sub, Hunts);
+    this.puzzleRefCounter = new RefCountedObserverMap(sub, Puzzles);
+
+    this.guessWatch = this.guessCursor.observeChanges({
+      added: (id, fields) => {
+        this.guesses[id] = { _id: id, ...fields } as GuessType;
+        this.huntRefCounter.incref(fields.hunt!);
+        this.puzzleRefCounter.incref(fields.puzzle!);
+        this.sub.added(Guesses.tableName, id, fields);
+      },
+
+      changed: (id, fields) => {
+        const huntUpdated = Object.prototype.hasOwnProperty.call(fields, 'hunt');
+        const puzzleUpdated = Object.prototype.hasOwnProperty.call(fields, 'puzzle');
+
+        // The order of operations here is important to avoid transient
+        // inconsistencies
+        if (huntUpdated) {
+          this.huntRefCounter.incref(fields.hunt!);
+        }
+        if (puzzleUpdated) {
+          this.puzzleRefCounter.incref(fields.puzzle!);
+        }
+        this.sub.changed(Guesses.tableName, id, fields);
+        if (huntUpdated) {
+          this.huntRefCounter.decref(this.guesses[id].hunt);
+        }
+        if (puzzleUpdated) {
+          this.puzzleRefCounter.decref(this.guesses[id].puzzle);
+        }
```

Comment:
I think after the decrefs are done, you should update your saved values in `this.guesses[id]` with the new field values?  Otherwise if there's another change, you'll be dropping refs on the original values, not the most-recent-predecessor values.  (Not like puzzle or hunt should ever be changing on guesses, but on principle...)

Reply:
Ah yeah, I had this line but lost it in a refactor. Pushed a fix.